### PR TITLE
Feature/Youtube API for Video [ED-1166]

### PR DESCRIPTION
**Pull request by kobizz**

### CODE REVIEW

File: assets/dev/js/frontend/handlers/video.js [hunk 1]
```diff
@@ -33,7 +33,12 @@ export default class Video extends elementorModules.frontend.handlers.Base {
 
 	playVideo() {
 		if ( this.elements.$video.length ) {
-			this.elements.$video[ 0 ].play();
+			// this.player exists only for YouTube videos, and its play function is different.
+			if ( this.player ) {
```
Comment (matipojo) [inline assets/dev/js/frontend/handlers/video.js:37]: Is there a better identifier? `player` may be a property used by other players, not only youtube.

File: assets/dev/js/frontend/handlers/video.js [hunk 2]
```diff
@@ -66,10 +71,79 @@ export default class Video extends elementorModules.frontend.handlers.Base {
...
+						}
+					},
+					onStateChange: ( event ) => {
+						if ( event.data === YT.PlayerState.ENDED && elementSettings.loop ) {
+							this.player.seekTo( elementSettings.start || 0 );
+						}
+					},
+				},
+				playerVars: {
+					controls: elementSettings.controls ? 1 : 0,
```
Comment (matipojo) [inline assets/dev/js/frontend/handlers/video.js:95]: Here and next vars
```suggestion
					controls: ! ! elementSettings.controls,
```

File: assets/dev/js/frontend/handlers/video.js [hunk 3]
```diff
@@ -66,10 +71,79 @@ export default class Video extends elementorModules.frontend.handlers.Base {
...
+					rel: elementSettings.rel ? 1 : 0,
+					playsinline: elementSettings.play_on_mobile ? 1 : 0,
+					modestbranding: elementSettings.modestbranding ? 1 : 0,
+					autoplay: elementSettings.autoplay ? 1 : 0,
+					start: elementSettings.start,
+					end: elementSettings.end,
```
Comment (matipojo) [inline assets/dev/js/frontend/handlers/video.js:101]: It's needed? it looks like the start/end are set in the next lines.

File: includes/widgets/video.php [hunk 4]
```diff
@@ -788,37 +797,42 @@ protected function render() {
...
-			$is_static_render_mode = Plugin::$instance->frontend->is_static_render_mode();
-			$post_id = get_queried_object_id();
+				$this->render_hosted_video();
 
-			if ( $is_static_render_mode ) {
-				$video_html = Embed::get_embed_thumbnail_html( $video_url, $post_id );
+				$video_html = ob_get_clean();
 			} else {
-				$video_html = Embed::get_embed_html( $video_url, $embed_params, $embed_options );
+				$is_static_render_mode = Plugin::$instance->frontend->is_static_render_mode();
```
Comment (matipojo) [inline includes/widgets/video.php:821]: The static render should work also for youtube videos.
It's recommended to create a Video_Base class that each provider can extend it.

File: includes/widgets/video.php [hunk 5]
```diff
@@ -873,7 +887,14 @@ protected function render() {
 						] );
 					}
 				} else {
-					$this->add_render_attribute( 'image-overlay', 'style', 'background-image: url(' . Group_Control_Image_Size::get_attachment_image_src( $settings['image_overlay']['id'], 'image_overlay', $settings ) . ');' );
+					// When there is an image URL but no ID, it means the overlay image is the placeholder. In this case, get the placeholder URL.
```
Comment (matipojo) [inline includes/widgets/video.php:890]: Not necessarily, in some cases, the URL is an external URL.

File: assets/dev/js/frontend/handlers/video.js [hunk 6]
```diff
@@ -66,10 +71,79 @@ export default class Video extends elementorModules.frontend.handlers.Base {
...
 	bindEvents() {
 		this.elements.$imageOverlay.on( 'click', this.handleVideo.bind( this ) );
 	}
 
+	onInit() {
+		super.onInit();
+
+		const elementSettings = this.getElementSettings();
+
+		if ( 'youtube' !== elementSettings.video_type ) {
```
Comment (matipojo) [inline assets/dev/js/frontend/handlers/video.js:130]: It's recommended to create a VideoBase class that each provider can extend it.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/frontend/handlers/video.js (hunk 2 above)
Comment (kobizz) [reply assets/dev/js/frontend/handlers/video.js:95]: According to [YouTube API](https://developers.google.com/youtube/player_parameters.html?playerVersion=HTML5)

File: includes/widgets/video.php (hunk 4 above)
Comment (kobizz) [reply includes/widgets/video.php:821]: Will be considered as a maintenance task

File: includes/widgets/video.php (hunk 5 above)
Comment (kobizz) [reply includes/widgets/video.php:890]: In which case?

File: assets/dev/js/frontend/handlers/video.js (hunk 6 above)
Comment (kobizz) [reply assets/dev/js/frontend/handlers/video.js:130]: Will be considered as a maintenance task

File: assets/dev/js/frontend/handlers/video.js (hunk 3 above)
Comment (udidol) [reply assets/dev/js/frontend/handlers/video.js:101]: Yes, this is actually needed and the next lines are not, will remove them.

File: assets/dev/js/frontend/handlers/video.js (hunk 1 above)
Comment (udidol) [reply assets/dev/js/frontend/handlers/video.js:37]: Done, Changed to `this.youtubePlayer`

File: assets/dev/js/frontend/handlers/video.js (hunk 3 above)
Comment (udidol) [reply assets/dev/js/frontend/handlers/video.js:101]: Done.

File: includes/widgets/video.php [hunk 7]
```diff
@@ -873,7 +887,14 @@ protected function render() {
...
+					if ( empty( $settings['image_overlay']['id'] && ! empty( $settings['image_overlay']['url'] ) ) ) {
+						$image_url = Utils::get_placeholder_image_src();
```
Comment (kobizz) [inline includes/widgets/video.php:892]:
```suggestion
						$image_url = $settings['image_overlay']['url'];
```

**Review by matipojo — APPROVED** [review]

**Pull request merged**